### PR TITLE
[SPARK-39294][SQL] Support vectorized Orc scans with DEFAULT values

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Support vectorized Orc scans when the table schema has associated DEFAULT column values.

(Note, this PR depends on https://github.com/apache/spark/pull/36672 which adds the same for Parquet files.)

Example:

```
create table t(i int) using orc;
insert into t values(42);
alter table t add column s string default concat('abc', def');
select * from t;
> 42, 'abcdef'
```

### Why are the changes needed?

This change makes it easier to build, query, and maintain tables backed by Orc data.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

This PR includes new test coverage.

### CODE REVIEW

**Comment by dtenedor** [comment]
@gengliangwang and/or @HyukjinKwon this is DEFAULT column support for the vectorized Orc reader, it depends on https://github.com/apache/spark/pull/36672 which adds the Parquet support.

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun and @viirya FYI

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @HyukjinKwon .

Could you rebase the PR in order to pass CI?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:278]: This PR targets Orc reader, why needing to change Parquet reader? This looks like newly added code, not moved code.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java:294]: If `isPresent` is false, throwing exception here, but in Orc reader it silently ignores that and continues to pull nulls. Why the difference?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java
Comment (dtenedor) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Yes, apologies for confusion, this PR is based off of https://github.com/apache/spark/pull/36672 which changes the Parquet reader. This PR only adds changes for the Orc reader on top of that. Please feel free to wait to review this PR again until the base PR is merged.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java
Comment (dtenedor) [reply sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java]: Good catch, I updated the behavior to be the same.

**Comment by gengliangwang** [comment]
@dtenedor The PR https://github.com/apache/spark/pull/36672/ is merged. Can you remove the Parquet-related code in this one? I will review this one after that today.

**Comment by dtenedor** [comment]
@gengliangwang sure, I just synced my Apache Spark repository fork to include the PR https://github.com/apache/spark/pull/36672 and then merged the changes into this branch. This PR now just shows changes to two files. Thanks for your help.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (if CI passed)

**Review by viirya — APPROVED** [review]

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

**Pull request closed without merging**
